### PR TITLE
clone_test.go: Remove use of RunVMAndExpectLaunchWithRunStrategy func

### PR DESCRIPTION
**Pull request by Barakmor1**
clone_test.go: Remove use of RunVMAndExpectLaunchWithRunStrategy func
RunVMAndExpectLaunchWithRunStrategy should be deprecated as it is part
of tests_utils.go and use expensive get and update API calls
    
Instead of Updating VM runStrategy in clone_test.go create vm with
with the desired runStrategy.

### What this PR does
Before this PR:

After this PR:

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/clone_test.go [hunk 1]
```diff
@@ -698,8 +700,24 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
 						libinstancetype.WaitForVMInstanceTypeRevisionNames(sourceVM.Name, virtClient)
 
 						if toRunSourceVM {
-							By("Starting the VM and expecting it to run")
-							sourceVM = RunVMAndExpectLaunchWithRunStrategy(virtClient, sourceVM, virtv1.RunStrategyAlways)
+							runStrategy := virtv1.RunStrategyAlways
+							sourceVM.Spec.RunStrategy = &runStrategy
```
Comment (nunnatsa) [inline tests/clone_test.go:704]: Please use `ptr.To()`
```suggestion
							sourceVM.Spec.RunStrategy = ptr.To(virtv1.RunStrategyAlways)
```

File: tests/clone_test.go [hunk 2]
```diff
@@ -698,8 +700,24 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
...
+							Expect(err).ToNot(HaveOccurred())
+							sourceVM, err = virtClient.VirtualMachine(sourceVM.Namespace).Patch(context.Background(), sourceVM.Name, types.JSONPatchType, payload, v1.PatchOptions{})
+							Expect(err).ToNot(HaveOccurred())
+
+							By("VM has the running condition")
+							Eventually(func() bool {
+								vm, err := virtClient.VirtualMachine(sourceVM.Namespace).Get(context.Background(), sourceVM.Name, v1.GetOptions{})
+								Expect(err).ToNot(HaveOccurred())
+								return vm.Status.Ready
+							}, 300*time.Second, 1*time.Second).Should(BeTrue())
```
Comment (nunnatsa) [inline tests/clone_test.go:720]: Please use `WithTimeout` and `WithPolling` instead of anonymous interval parameters.

**Review by nunnatsa — COMMENTED** [review]
Added a few inline comments

File: tests/clone_test.go (hunk 2 above)
Comment (nunnatsa) [reply tests/clone_test.go:720]:
```suggestion
							}).WithTimeout(300*time.Second).WithPolling(1*time.Second).Should(BeTrue())
```

File: tests/clone_test.go (hunk 1 above)
Comment (Barakmor1) [reply tests/clone_test.go]: Done :)

File: tests/clone_test.go (hunk 2 above)
Comment (Barakmor1) [reply tests/clone_test.go]: Done :)

File: tests/clone_test.go [hunk 3]
```diff
@@ -696,8 +699,23 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
 						libinstancetype.WaitForVMInstanceTypeRevisionNames(sourceVM.Name, virtClient)
 
 						if toRunSourceVM {
-							By("Starting the VM and expecting it to run")
-							sourceVM = RunVMAndExpectLaunchWithRunStrategy(virtClient, sourceVM, virtv1.RunStrategyAlways)
+							sourceVM.Spec.RunStrategy = ptr.To(virtv1.RunStrategyAlways)
```
Comment (ShellyKa13) [inline tests/clone_test.go:702]: why not extract it to a function that can be used everywhere that calls this RunVMAndExpectLaunchWithRunStrategy function?

File: tests/clone_test.go [hunk 4]
```diff
@@ -696,8 +699,23 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
...
+								patch.WithAdd("/spec/runStrategy", virtv1.RunStrategyAlways),
+								patch.WithAdd("/spec/running", nil),
+							)
+							payload, err := patchSet.GeneratePayload()
+							Expect(err).ToNot(HaveOccurred())
+							sourceVM, err = virtClient.VirtualMachine(sourceVM.Namespace).Patch(context.Background(), sourceVM.Name, types.JSONPatchType, payload, v1.PatchOptions{})
+							Expect(err).ToNot(HaveOccurred())
+
+							By("VM has the running condition")
+							Eventually(func() bool {
```
Comment (ShellyKa13) [inline tests/clone_test.go:714]: can use:
Eventually(ThisVM(vm), 360*time.Second, 1*time.Second).Should(BeReady())

File: tests/clone_test.go (hunk 3 above)
Comment (Barakmor1) [reply tests/clone_test.go:702]: Where would you put this function?

File: tests/clone_test.go (hunk 4 above)
Comment (Barakmor1) [reply tests/clone_test.go]: Oh nice i didn't know about these functions :)

Done

File: tests/clone_test.go (hunk 3 above)
Comment (akalenyu) [reply tests/clone_test.go:702]: Maybe we want an equivalent of https://github.com/kubevirt/kubevirt/blob/53f1c51472840a3e30a151392ff0b0a1737a4f2f/tests/libkvconfig/kvconfig.go#L37-L49 for VMs?

Comment (akalenyu) [reply tests/clone_test.go:702]: If this sounds like overkill we can just change this test to create a VM that already has the correct runstrategy upfront

Comment (enp0s3) [reply tests/clone_test.go:702]: @akalenyu Sounds good. I think this kind of VM-patch function can live in the future `libclient` package.

**Review by enp0s3 — COMMENTED** [review]
@Barakmor1 Thanks for the PR! I think it worth waiting for the refactoring of all the `RunVM*` functions, extracting them to a dedicated package. I think that if RunVMAndExpectLaunchWithRunStrategy wouldn't be in utils.go then its legitimate to to use it here.

**Comment by Barakmor1** [comment]
> @Barakmor1 Thanks for the PR! I think it worth waiting for the refactoring of all the `RunVM*` functions, extracting them to a dedicated package. I think that if RunVMAndExpectLaunchWithRunStrategy wouldn't be in utils.go then its legitimate to to use it here.

Maybe :)
We realized that we don't actually need this function call in these tests.

**Review by enp0s3 — COMMENTED** [review]
From what I see in the API:
```
	// Running state indicates the requested running state of the VirtualMachineInstance
	// mutually exclusive with Running
	RunStrategy *VirtualMachineRunStrategy `json:"runStrategy,omitempty" optional:"true"`
```
It seems like it is enough to use only `spec.Running`

**Comment by Barakmor1** [comment]
> From what I see in the API:
> 
> ```
> 	// Running state indicates the requested running state of the VirtualMachineInstance
> 	// mutually exclusive with Running
> 	RunStrategy *VirtualMachineRunStrategy `json:"runStrategy,omitempty" optional:"true"`
> ```
> 
> It seems like it is enough to use only `spec.Running`

Yes you are absolutely right, thanks! :)

**Comment by iholder101** [comment]
> From what I see in the API:
> 
> ```
> 	// Running state indicates the requested running state of the VirtualMachineInstance
> 	// mutually exclusive with Running
> 	RunStrategy *VirtualMachineRunStrategy `json:"runStrategy,omitempty" optional:"true"`
> ```
> 
> It seems like it is enough to use only `spec.Running`

Nice!
But please use run strategy instead of spec.running which needs to be deprecated.

**Comment by iholder101** [comment]
Thanks @Barakmor1!
/approve

**Review by dankenigsberg — APPROVED** [review]

**Comment by Barakmor1** [comment]
/hold

As much as I want to get this in, we still need to wait for @akalenyu  to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.

**Comment by akalenyu** [comment]
> /hold
> 
> As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.

I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR

**Comment by Barakmor1** [comment]
> > /hold
> > As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.
> 
> I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR

I’ve noticed that the `pull-kubevirt-e2e-k8s-1.28-sig-storage`, `pull-kubevirt-e2e-k8s-1.29-sig-storage`, and `pull-kubevirt-e2e-k8s-1.30-sig-storage` lanes have been consistently failing in this PR with the error "Running and RunStrategy are mutually exclusive" due to the [Running setup in libvmi](https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/libvmi/vm.go#L44).
I got the impression that you might be planning to remove this and set a runStrategy instead. 

Could you please let me know if I’ve understood this correctly?

Thank you!

**Comment by akalenyu** [comment]
> > > /hold
> > > As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.
> > 
> > 
> > I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR
> 
> I’ve noticed that the `pull-kubevirt-e2e-k8s-1.28-sig-storage`, `pull-kubevirt-e2e-k8s-1.29-sig-storage`, and `pull-kubevirt-e2e-k8s-1.30-sig-storage` lanes have been consistently failing in this PR with the error "Running and RunStrategy are mutually exclusive" due to the [Running setup in libvmi](https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/libvmi/vm.go#L44). I got the impression that you might be planning to remove this and set a runStrategy instead.
> 
> Could you please let me know if I’ve understood this correctly?
> 
> Thank you!

Right, if you use running instead of runstrategy, this won't fail and the PR can move ahead

**Comment by Barakmor1** [comment]
> > > > /hold
> > > > As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.
> > > 
> > > 
> > > I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR
> > 
> > 
> > I’ve noticed that the `pull-kubevirt-e2e-k8s-1.28-sig-storage`, `pull-kubevirt-e2e-k8s-1.29-sig-storage`, and `pull-kubevirt-e2e-k8s-1.30-sig-storage` lanes have been consistently failing in this PR with the error "Running and RunStrategy are mutually exclusive" due to the [Running setup in libvmi](https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/libvmi/vm.go#L44). I got the impression that you might be planning to remove this and set a runStrategy instead.
> > Could you please let me know if I’ve understood this correctly?
> > Thank you!
> 
> Right, if you use running instead of runstrategy, this won't fail and the PR can move ahead

Yes, but when we spoke offline, we decided that "running" should not be used here because we need to deprecate it. :)

**Comment by Barakmor1** [comment]
> > > > > /hold
> > > > > As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.
> > > > 
> > > > 
> > > > I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR
> > > 
> > > 
> > > I’ve noticed that the `pull-kubevirt-e2e-k8s-1.28-sig-storage`, `pull-kubevirt-e2e-k8s-1.29-sig-storage`, and `pull-kubevirt-e2e-k8s-1.30-sig-storage` lanes have been consistently failing in this PR with the error "Running and RunStrategy are mutually exclusive" due to the [Running setup in libvmi](https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/libvmi/vm.go#L44). I got the impression that you might be planning to remove this and set a runStrategy instead.
> > > Could you please let me know if I’ve understood this correctly?
> > > Thank you!
> > 
> > 
> > Right, if you use running instead of runstrategy, this won't fail and the PR can move ahead
> 
> Yes, but when we spoke offline, we decided that "running" should not be used here because we need to deprecate it. :)

/cc @EdDev @dankenigsberg WDYT Should we wait for a fix in libvmi or just use `running` for now?

File: tests/clone_test.go [hunk 5]
```diff
@@ -682,24 +682,19 @@ var _ = Describe("[Serial]VirtualMachineClone Tests", Serial, func() {
 							Name: preference.Name,
 							Kind: "VirtualMachinePreference",
 						}
+					})
 
+					DescribeTable("should create new ControllerRevisions for cloned VM", Label("instancetype", "clone"), func(runStrategy virtv1.VirtualMachineRunStrategy) {
+						sourceVM.Spec.RunStrategy = &runStrategy
```
Comment (EdDev) [inline tests/clone_test.go:688]: If I understand correctly, the current problem is that `Running` and `RunStrategy` cannot co-exist.
We prefer `RunStrategy` and we should stop using `Running` (and even deprecate it).

Therefore, I see you have these two options:
- Fix the problem from the source, i.e. `libvmi`.
- Fix the problem in the most appropiate location for this specific test context: The `BeforeEach` that creates the VM. This means to add there `sourceVM.Spec.Running = nil`.

If fixing the original problem is causing an avalanche of changes all over the codebase, then you can go with the second option for now.

I want to emphasize that it is more important to identify and fix root causes over finishing a specific small cleanup.
It is very common to find bigger problems while working on small scoped changes.

Comment (Barakmor1) [reply tests/clone_test.go:688]: Sounds good to me.

**Comment by akalenyu** [comment]
> > > > > /hold
> > > > > As much as I want to get this in, we still need to wait for @akalenyu to remove the Running configurations from libvmi. Otherwise, the tests will constantly fail.
> > > > 
> > > > 
> > > > I don't think we should block this PR over this - instead, it'll just become one of the refactoring targets in a subsequent PR
> > > 
> > > 
> > > I’ve noticed that the `pull-kubevirt-e2e-k8s-1.28-sig-storage`, `pull-kubevirt-e2e-k8s-1.29-sig-storage`, and `pull-kubevirt-e2e-k8s-1.30-sig-storage` lanes have been consistently failing in this PR with the error "Running and RunStrategy are mutually exclusive" due to the [Running setup in libvmi](https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/libvmi/vm.go#L44). I got the impression that you might be planning to remove this and set a runStrategy instead.
> > > Could you please let me know if I’ve understood this correctly?
> > > Thank you!
> > 
> > 
> > Right, if you use running instead of runstrategy, this won't fail and the PR can move ahead
> 
> Yes, but when we spoke offline, we decided that "running" should not be used here because we need to deprecate it. :)

Yes, you can mark it with a TODO for now, but it definitely shouldn't block the PR as we haven't even gotten close to deprecating Running. Note we will have testing keep using Running in some places since we can't stop testing it altogether immediately

File: tests/clone_test.go (hunk 5 above)
Comment (Barakmor1) [reply tests/clone_test.go:688]: > I want to emphasize that it is more important to identify and fix root causes over finishing a specific small cleanup.

AFAIK, @iholder101 already [opened an issue](https://github.com/kubevirt/kubevirt/issues/11993) about deprecating the Running field.

**Review by ShellyKa13 — APPROVED** [review]
/lgtm

**Pull request merged**